### PR TITLE
LibLine: Add styling; LibJS/JS repl: add syntax highlighting

### PR DESCRIPTION
**Pull request by alimpfard**
Perhaps best served as separate PRs, but they were on the same branch...

These patches mainly add live styling to LibLine, and are my second take on the syntax-highlighted js repl.
If this goes well, we might even be able to provide optional highlighting for the shell itself~

### CODE REVIEW

File: Libraries/LibLine/Editor.h [hunk 1]
```diff
@@ -78,6 +81,7 @@ class Editor {
 
     Function<Vector<String>(const String&)> on_tab_complete_first_token = nullptr;
     Function<Vector<String>(const String&)> on_tab_complete_other_token = nullptr;
+    Function<void(Editor&)> on_display_refresh = nullptr;
```
Comment (awesomekling) [inline Libraries/LibLine/Editor.h:84]: No need to explicitly initialize `Function` objects with `nullptr`, as they will default-initialize the same way.

File: Libraries/LibLine/Editor.h [hunk 2]
```diff
@@ -100,9 +111,37 @@ class Editor {
     void vt_save_cursor();
     void vt_restore_cursor();
     void vt_clear_to_end_of_line();
+    void vt_clear_lines(size_t count_above, size_t count_below = 0);
+    void vt_move_relative(int x, int y);
+    void vt_apply_style(const Style&);
+
+    Style find_applicable_style(size_t offset);
```
Comment (awesomekling) [inline Libraries/LibLine/Editor.h:118]: This does not mutate state, so should be `const` qualified.

File: Libraries/LibLine/Span.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+namespace Line {
+class Span {
+public:
+    Span(size_t start, size_t end)
+        : m_beginning(start)
+        , m_end(end)
+    {
+    }
+
+    size_t begining() const { return m_beginning; }
```
Comment (awesomekling) [inline Libraries/LibLine/Span.h:39]: Typo here:
```suggestion
    size_t beginning() const { return m_beginning; }
```

File: Userland/js.cpp [hunk 4]
```diff
@@ -351,6 +353,126 @@ int main(int argc, char** argv)
 
         editor = make<Line::Editor>();
         editor->initialize();
+        if (syntax_highlight)
+            editor->on_display_refresh = [](Line::Editor& editor) {
+                using namespace Line;
+                using Color = Style::Color;
```
Comment (awesomekling) [inline Userland/js.cpp:359]: ew but ok

Comment (alimpfard) [reply Userland/js.cpp:359]: honestly...`Line::Style::Color::Red` is a mouthful and a half...do you have any alternatives to this by chance?
`Line::Color` sounds...weird

Comment (alimpfard) [reply Userland/js.cpp:359]: I'm gonna just leave it be as `Line::Style::Color` and...maybe some day someone will find a better name.

File: Libraries/LibLine/Style.h [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        }
+        Color m_color;
+    };
+
+    static constexpr UnderlineTag Underline {};
+    static constexpr BoldTag Bold {};
+    static constexpr ItalicTag Italic {};
+
+    // prepare for the horror of templates
+    template <typename T, typename... Rest>
```
Comment (alimpfard) [inline Libraries/LibLine/Style.h:81]: @awesomekling now I understand why you can't remember where these dots go lol

**Comment by alimpfard** [comment]
The sheer number of rebases I've had to do on this one is staggering.

I am not touching this js repl again, it's too fast-moving for my tastes

**Comment by awesomekling** [comment]
All right let's merge it :D

**Pull request merged**